### PR TITLE
[Brave News]: Build feed off the main thread

### PR DESCRIPTION
**Pull request by fallaciousreasoning**
Resolves https://github.com/brave/brave-browser/issues/35695

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

1. In a new profile click the `Learn More` link on the Enable Brave News popup
2. Navigate back
3. The page should not crash
4. Enable Brave News
5. The publishers should load
6. The feed should load
7. Disable Brave news
8. Enable Brave news
9. THe publishers should load 
10. THe Feed should load
11. Disable Brave news, refresh the page
12. Enable Brave news
13. The publishers should load
14. The feed should load

### CODE REVIEW

**Comment by petemill** [comment]
Looks like we definitely don't need this part, but also seems like it doesn't remove the ~startup requests for data since the TabHelper (created for every Tab's WebContents) ensures the data has started to be fetched in its ctor. And this data _is_ needed for the feature to work. So, unless we're going to delay using the feature away from the 1st WebContents, then we should probably also make sure we're doing all we can to not block the main thread when requesting, parsing or processing the publisher or feed data.

File: components/brave_news/browser/brave_news_controller.cc
Comment (petemill) [inline components/brave_news/browser/brave_news_controller.cc:782]: we should also verify Android is still ok without this part. I know CI on android just failed but it just seems like an unrelated transient abort. Just noting that hopefully Android's settings (or feed) will gracefully handle having no data and then getting data asynchronously.

Comment (fallaciousreasoning) [reply components/brave_news/browser/brave_news_controller.cc:782]: Ah, that's a good point - I'm going to revert this bit so we can land the main change more easily.

**Review by petemill — APPROVED** [review]
Seems sensible to remove the prefetch and do that separately if the thread moving is easier to reason

File: components/brave_news/browser/feed_v2_builder.cc
Comment (sangwoo108) [inline components/brave_news/browser/feed_v2_builder.cc:1165]: optional: It'd be good to be consistent about naming

```suggestion
                                 FeedGenerator generator,
```

File: components/brave_news/browser/feed_v2_builder.h
Comment (sangwoo108) [inline components/brave_news/browser/feed_v2_builder.h:92]: optional: can we add one more comment, like

```suggestion
  // data in |builder| will not be changed or invalidated while the generator is running. 
```

so as to let others know that it's safe to have `const&` rather than `WeakPtr`.

File: components/brave_news/browser/feed_v2_builder.cc
Comment (sangwoo108) [inline components/brave_news/browser/feed_v2_builder.cc:1259]: Regardless of logical invariant of the builder's state,
to me it feel like accessing builder's member variable and publisher_controller is exposing us to potential risk of thread safety.  If we're to do this, PublisherController's GetLastLocale() and others should be guaranteed of thread safety. 

I think bind all recipes like publisher ids, locale, last publishers, signal from the main thread would be much safer.

(GenerateAllfeed as well)

Comment (fallaciousreasoning) [reply components/brave_news/browser/feed_v2_builder.cc:1259]: Okay cool, I was considering that but got scared off by the fact that we were doing this for perf. Honestly, it'd make me feel a lot better about this to do it that way :+1:

**Comment by fallaciousreasoning** [comment]
@sangwoo108 thanks for the feedback - I'm going to copy everything into the lambda. I've got some compile errors, so going to push tomorrow :smile:

File: components/brave_news/browser/feed_v2_builder.cc
Comment (sangwoo108) [reply components/brave_news/browser/feed_v2_builder.cc:1259]: > got scared off by the fact that we were doing this for perf

Yeah, now I see where you're coming from. Your concern makes sense if all vector is huge. I think we can try using RefCountedThreadSafe and others. Let me try a bit.

File: components/brave_news/browser/feed_v2_builder.cc
Comment (atuchin-m) [inline components/brave_news/browser/feed_v2_builder.cc:1194]: 1. `base::SequencedTaskRunner::GetCurrentDefault()` referrers to UI thread here. I suggest use https://chromium.googlesource.com/chromium/src/+/main/docs/threading_and_tasks.md#direct-posting-to-the-thread-pool to post to the thread pool.
2. You [can't dereference WeakPtr](https://source.chromium.org/chromium/chromium/src/+/main:base/memory/weak_ptr.h;l=52-54) on different threads.
3. The posted task has a connection with `this`, so there is extremely high chance to create a data race. Even `GenerateAllFeed` uses `const FeedV2Builder&`, the things can be modified on UI. Also the method use some things from UI thread (e.g. PerfService **here**)
4. Could you please add `DCHECK(!content::BrowserThread::CurrentlyOn(content::BrowserThread::UI));` to processing function to verify it isn't UI?

File: components/brave_news/browser/feed_v2_builder.cc
Comment (atuchin-m) [inline components/brave_news/browser/feed_v2_builder.cc:1301]: `prefs_` is UI-only thing, can't be used on other threads.

File: components/brave_news/browser/feed_v2_builder.cc
Comment (atuchin-m) [reply components/brave_news/browser/feed_v2_builder.cc:1259]: I suggest not use `RefCountedThreadSafe`, it makes the code much harder to maintain in terms of avoiding data races.

There are a few alternatives:
1. Post everything each time you need as extra arguments as a copy in function/static method/lambda (depends on the data, could be costly).
2. Make a task runner inner class with necessary data. See FooService here: https://source.chromium.org/chromium/chromium/src/+/main:docs/threading_and_tasks_testing.md?q=OnTaskRunnerDeleter%20lang:md

File: components/brave_news/browser/feed_v2_builder.cc
Comment (atuchin-m) [reply components/brave_news/browser/feed_v2_builder.cc:1194]: Double checked in the debugger, it's still UI thread:
image

**Comment by atuchin-m** [comment]
Does the PR cover all the heavy request processing or the other usage will be fixed separately?
E.g. https://github.com/brave/brave-core/blob/master/components/brave_news/browser/publishers_parsing.cc#L24 (is called during startup or other parsing things.

File: components/brave_news/browser/feed_v2_builder.cc
Comment (sangwoo108) [reply components/brave_news/browser/feed_v2_builder.cc:1194]: Yeah, I realized that while I was trying to refactor. I believe we should use ThreadPool.

**Comment by atuchin-m** [comment]
Also it's not clear about `FeedV2Builder` lifetime and about safety of `base::Unretained` usage like this one:
https://github.com/brave/brave-core/blob/master/components/brave_news/browser/feed_v2_builder.cc#L934

After this change the callbacks start to be invoked async => if the object is destroyed we get UAF.

File: components/brave_news/browser/feed_v2_builder.h
Comment (sangwoo108) [inline components/brave_news/browser/feed_v2_builder.h:38]: I think exposing anonymous namespace in header isn't better than private inner class, as it makes other fils that include this header have the same declaration in anonymous namespace

File: components/brave_news/browser/feed_v2_builder.cc
Comment (sangwoo108) [inline components/brave_news/browser/feed_v2_builder.cc:1415]: nit:

```suggestion
            // can be quite slow. It's safe because all data |generator| requires is copied into the lambda.
```

**Review by sangwoo108 — APPROVED** [review]
LGTM! @atuchin-m , I think this PR is safe now. Could you make the final call? I think we can apply scoped_ref partially on problematic data later, if copying cost turns out to be too expensive.

File: components/brave_news/browser/feed_v2_builder.cc
Comment (atuchin-m) [inline components/brave_news/browser/feed_v2_builder.cc:1374]: `thread_checker_` is invalid here. We are post-tasking to any thread in pool, so the different tasks could use different threads.
Let's just check !UI.

File: components/brave_news/browser/feed_v2_builder.cc
Comment (atuchin-m) [inline components/brave_news/browser/feed_v2_builder.cc:1408]: I wonder what it's the typical size of `FeedGenerationInfo` and how much CPU we need to construct it (including all the for-loops here.

File: components/brave_news/browser/feed_v2_builder.cc
Comment (atuchin-m) [inline components/brave_news/browser/feed_v2_builder.cc:66]: Frankly, I missed where is the original `base::Value` from https://brave-today-cdn.brave.com/brave-today/feed.en_USjson that we got from DecodeService.

It seems it's already been processed on UI thread and been put to `this` as `topics_`, etc.
Because the list is huge (3500+ items) even just per-item copying it could result UI freeze..

Any options to completely process the huge data on non UI thread?

Comment (atuchin-m) [reply components/brave_news/browser/feed_v2_builder.cc:66]: Even on `master` with disabled `FeedV2Builder::GenerateFeed` the UI task takes 100ms on win-Static build ;(
It means that this PR only a part of a story. Also https://github.com/brave/brave-core/pull/21833#issuecomment-1921029106 should be taken into account.
image

image

File: components/brave_news/browser/feed_v2_builder.h
Comment (fallaciousreasoning) [reply components/brave_news/browser/feed_v2_builder.h:38]: Done!

File: components/brave_news/browser/feed_v2_builder.cc
Comment (fallaciousreasoning) [reply components/brave_news/browser/feed_v2_builder.cc:1408]: I imagine the typical size is large. 3500 `FeedItems`, similar number of `TopicArticles`, and maybe ~250 channels/publishers.

That said, I just timed this on my machine and it takes 11ms for this bit of code, which constructs the `FeedGenerationInfo`:
image

File: components/brave_news/browser/feed_v2_builder.cc
Comment (fallaciousreasoning) [reply components/brave_news/browser/feed_v2_builder.cc:66]: Agreed that we should push everything else into a background thread - that said, currently the `PublisherParsing` function you linked to seems to take 2ms, so I don't think it's urgent.

After a bit of investigation, I think the other 100ms are coming from the `FeedFetcher`. I'm going to try and push that into a background thread too.

File: components/api_request_helper/api_request_helper.h
Comment (fallaciousreasoning) [inline components/api_request_helper/api_request_helper.h:61]: Added this so I can move `value_body` to another thread, rather than cloning it.

File: components/brave_news/browser/feed_fetcher.cc
Comment (fallaciousreasoning) [inline components/brave_news/browser/feed_fetcher.cc:194]: I think this was the main factor in the performance issues - when I was looking at a trace most of the time was spent comparing `GURL`s

Comment (sangwoo108) [reply components/brave_news/browser/feed_fetcher.cc:194]: the operator==() overload for GURL is using string::operator==. So I guess there wouldn't be big difference. https://source.chromium.org/chromium/chromium/src/+/44812c2055259e0e48f5077280d918ab64ab390b:url/gurl.cc;l=544

Comment (fallaciousreasoning) [reply components/brave_news/browser/feed_fetcher.cc:194]: I was thinking that maybe it does `.spec()` for every comparison, so in `flat_set` the worst case would be `O(n)` calls to `spec()`

Maybe using `std::set` would be better here?

Comment (sangwoo108) [reply components/brave_news/browser/feed_fetcher.cc:194]: Yeah, flat_set is designed to be used for handful data. [As per the document](https://source.chromium.org/chromium/chromium/src/+/3e815ef723d2b1a4b467fff93e16a177ebae1626:base/containers/README.md;l=77), it's O(n) on insertion and deletion. 

And std::set is implemented as RB tree, so it's performance would be O(logn) in the worst case. so std::unordered_set, which has O(1) complexity, could be even better

Comment (fallaciousreasoning) [reply components/brave_news/browser/feed_fetcher.cc]: Done! I'm doing `.reserve(n)` now too - as I understand it'll make `std::unordered_set` reserve enough space to fit `n` items at the `load_factor` of the set.

Comment (sangwoo108) [reply components/brave_news/browser/feed_fetcher.cc]: Wow, didn't know about it. Nice catch! It seems that reserving would help us to avoid rehashing everything 👍

File: components/brave_news/browser/feed_v2_builder.cc
Comment (atuchin-m) [reply components/brave_news/browser/feed_v2_builder.cc:1415]: Sadly, 11ms also sounds like an issue for UI thread, especially for Android. UI thread is in charge of making frames each 33ms on average, therefore heavy computation is prohibited here.
From [the upstream doc](https://source.chromium.org/chromium/chromium/src/+/main:docs/threading_and_tasks.md): 
```
Do not perform expensive computation or blocking IO on the main thread
(a.k.a. “UI” thread in the browser process) or IO thread (each process's thread for receiving IPC).
A busy UI / IO thread can cause user-visible latency, so prefer running that work on the thread pool.
```

I believe the only way to solve this issue completely is to move all the inner things to a dedicated sequence (aka virtual thread in the pool).
[A good example](https://source.chromium.org/chromium/chromium/src/+/main:components/subresource_filter/content/browser/async_document_subresource_filter.h;l=188-189?q=async_document_subresource_filter.h) in Chromium code.
Otherwise, it's about impossible to manipulate such large structures on UI thread.

I'm not against solving the issue step-by-step, but I'd like to discuss and sync the final plan before we going to merge an individual PRs

File: components/brave_news/browser/feed_fetcher.cc
Comment (atuchin-m) [reply components/brave_news/browser/feed_fetcher.cc]: for 3000+ GURL items `unordered_set` looks like a best choice.
BTW, it's not clear for me why the data URLs used as a key here (not something like id or hash).

File: components/api_request_helper/api_request_helper.h
Comment (atuchin-m) [reply components/api_request_helper/api_request_helper.h:61]: Yep, it's also an important thing.
I have a PR to fix that problem more globally: https://github.com/brave/brave-core/pull/21876/files#diff-ecc8fbcb6cb29a6a68559a53dbfe02bdf099789e5df43bfef140084cb010c156

File: components/brave_news/browser/feed_fetcher.cc
Comment (atuchin-m) [reply components/brave_news/browser/feed_fetcher.cc]: hm, url_hash also a super long string ;(
image

File: components/brave_news/browser/feed_v2_builder.cc
Comment (atuchin-m) [reply components/brave_news/browser/feed_v2_builder.cc]: Thanks for fixing `FeedFetcher`.
If the parsed data (std::vectors) was small enough we would stop here.  Unfortunately, it seems even manipulating the parsed data on UI thread could be a problem: https://github.com/brave/brave-core/pull/21833#discussion_r1477747653

File: components/brave_news/browser/feed_fetcher.cc
Comment (atuchin-m) [inline components/brave_news/browser/feed_fetcher.cc:171]: Is it safe to call `callback` later?

File: components/brave_news/browser/feed_fetcher.cc
Comment (atuchin-m) [inline components/brave_news/browser/feed_fetcher.cc:229]: The same question about `callback` safety.

File: components/brave_news/browser/feed_fetcher.cc
Comment (fallaciousreasoning) [reply components/brave_news/browser/feed_fetcher.cc]: So the reason I wasn't using `url_hash` is that in the past the backend has been very inconsistent in providing it so it hasn't been safe to rely on.

Unfortunately, that makes url the closest thing to a guaranteed Id we have.

File: components/brave_news/browser/feed_fetcher.cc
Comment (fallaciousreasoning) [reply components/brave_news/browser/feed_fetcher.cc]: I've updated these to check a weakptr before running the callback.

File: components/brave_news/browser/feed_fetcher.cc
Comment (fallaciousreasoning) [reply components/brave_news/browser/feed_fetcher.cc]: Yeah, thinking about it, previously the `FeedFetcher` wouldn't fire the callback if it had been destroyed, and we might be relying on that. Nice catch!

File: components/api_request_helper/api_request_helper.h
Comment (fallaciousreasoning) [reply components/api_request_helper/api_request_helper.h:61]: Sweet, that should help things heaps :smile: Lazily getting the string should help quite a bit.

File: components/brave_news/browser/feed_v2_builder.cc
Comment (fallaciousreasoning) [reply components/brave_news/browser/feed_v2_builder.cc:1415]: Yeah agreed, we should fix the root cause (that we're on the UI thread) rather than just putting bandaids over things. However, I think it's going to be a much more significant refactor to move everything into a sequence.

What I'm roughly thinking at the moment is this:
1. We land & uplift this PR, roughly as is.
2. We refactor `BraveNewsController` to have it's own sequence. Every call to subservices should happen on this sequence (this is going to be complicated because things are quite tightly tied to `PrefService` at the moment).
3. We change this `FeedGenerationInfo` to just take a reference (still handy to tie everything together) and get rid of the `base::ThreadPool` usages here.

Does that sound reasonable? Happy to jump on a call to discuss :smile:

Comment (fallaciousreasoning) [reply components/brave_news/browser/feed_v2_builder.cc:1415]: Okay, after a bit more research, I think something a bit like what `PasswordStore` does is probably the way to go.

We introduce a `BraveNewsControllerFrontEnd`. Lives on the UI thread, has a `scoped_refptr` to a `BraveNewsController` and `SequencedTaskRunner` which is posts tasks to the `BraveNewsController` on.

File: components/brave_news/browser/feed_fetcher.cc
Comment (atuchin-m) [inline components/brave_news/browser/feed_fetcher.cc:233]: Checking `fetcher` makes sense. Although, the lifetime of the callback could be different.
Frankly, I can't 100% sure that we don't add new crashes because posting the callback.

The code use a very callback-based approach. It's really hard to surely find all the entry points and make a conclusion that it's still safe..

File: components/brave_news/browser/feed_v2_builder.cc
Comment (atuchin-m) [reply components/brave_news/browser/feed_v2_builder.cc:1415]: Let's merge this PR (after we sure that all the callback are good), this should improve the situation.
The next steps are to solve the issue completely. I hope I could provide you some perf metrics to track the progress (in https://github.com/brave/brave-core/pull/21876)

File: components/brave_news/browser/feed_fetcher.cc
Comment (atuchin-m) [reply components/brave_news/browser/feed_fetcher.cc:233]: I other words, we changed the public contract of the class. Before we run all the callback sync, now we could postpone. A callback could be bind not only to `FeedFetcher` but to mojo-related interface (to send the answer) or something else. Better to double check all the users. Or even make renaming `DoSomething(..)` => `DoSomethingAsync(..)`

File: components/brave_news/browser/feed_v2_builder.cc
Comment (atuchin-m) [inline components/brave_news/browser/feed_v2_builder.cc:1349]: Looks a bit weired to me..
1. The method named `NotifyUpdateCompleted` but the actual notifications seems to be postoned to `on_notify_completed`.
2. We modify some class field like an update is finished but in fact we are still waiting to the callbacks.

File: components/brave_news/browser/feed_fetcher.cc
Comment (fallaciousreasoning) [reply components/brave_news/browser/feed_fetcher.cc:233]: I don't think we have changed the the public contract of the class - the callback was never invoked synchronously, only ever after all the combined feeds and direct feeds are fetched (quite apart from the network request, JSON parsing of the combined feed potentially happens in another process).

I added the check that `fetcher` wasn't destroyed because it was the only place I could think of where the contract was changing (previously the callback wouldn't be invoked if the `FeedFetcher` was destroyed, and the WeakPtr check maintains that guarantee).

File: components/brave_news/browser/feed_fetcher.cc
Comment (fallaciousreasoning) [reply components/brave_news/browser/feed_fetcher.cc:185]: Converted to a static function because `FeedSourceResult` is a private class.

File: components/brave_news/browser/feed_v2_builder.cc
Comment (fallaciousreasoning) [reply components/brave_news/browser/feed_v2_builder.cc:1349]: Ah nice catch - we don't actually need this anymore. It was to make sure the data wasn't changed while we were running callbacks.

However, because we take a copy of the data now we don't need this.

Also noticed a bug where |next_updates_| wouldn't get triggered :fearful: which I fixed.

**Review by atuchin-m — APPROVED** [review]

**Review by mkarolin — APPROVED** [review]
`DEPS`++

File: components/brave_news/browser/feed_v2_builder.cc
Comment (bridiver) [inline components/brave_news/browser/feed_v2_builder.cc:899]: it's not good to add content dependencies to components if they aren't strictly necessary because it makes this code unusable on ios

Comment (bridiver) [reply components/brave_news/browser/feed_v2_builder.cc:899]: It doesn't work here in a static method, but normally you can get the current sequence in the constructor and use that for sequence checks.

**Pull request merged**
